planner/core/find_best_task.go

```diff
@@ -814,52 +815,50 @@ func (is *PhysicalIndexScan) indexScanRowSize(idx *model.IndexInfo, ds *DataSour
-			indexCols = append(indexCols, idxExprCols[i])
-		} else {
-			// TODO: try to reuse the col generated when building the DataSource.
-			indexCols = append(indexCols, &expression.Column{
-				ID:       is.Table.Columns[idx.Columns[i].Offset].ID,
-				RetType:  &is.Table.Columns[idx.Columns[i].Offset].FieldType,
-				UniqueID: is.ctx.GetSessionVars().AllocPlanColumnID(),
-			})
-		}
-	}
-	setHandle := len(indexCols) > len(idx.Columns)
-	if !setHandle {
-		for i, col := range is.Columns {
-			if (mysql.HasPriKeyFlag(col.Flag) && is.Table.PKIsHandle) || col.ID == model.ExtraHandleID {
-				indexCols = append(indexCols, is.dataSourceSchema.Columns[i])
-				setHandle = true
-				break
+	is.NeedCommonHandle = is.Table.IsCommonHandle
+
+	if !is.NeedCommonHandle {
+		for i := len(is.IdxCols); i < len(idx.Columns); i++ {
+			if idxExprCols[i] != nil {
+				indexCols = append(indexCols, idxExprCols[i])
+			} else {
+				// TODO: try to reuse the col generated when building the DataSource.
+				indexCols = append(indexCols, &expression.Column{
+					ID:       is.Table.Columns[idx.Columns[i].Offset].ID,
+					RetType:  &is.Table.Columns[idx.Columns[i].Offset].FieldType,
+					UniqueID: is.ctx.GetSessionVars().AllocPlanColumnID(),
+				})
 			}
 		}
-	}
-
-	if is.Table.IsCommonHandle {
-		pkIdx := tables.FindPrimaryIndex(is.Table)
-		for _, col := range pkIdx.Columns {
-			indexCols = append(indexCols, &expression.Column{
-				ID:       is.Table.Columns[col.Offset].ID,
-				RetType:  &is.Table.Columns[col.Offset].FieldType,
-				UniqueID: is.ctx.GetSessionVars().AllocPlanColumnID(),
-			})
+		setHandle := len(indexCols) > len(idx.Columns)
+		if !setHandle {
+			for i, col := range is.Columns {
+				if (mysql.HasPriKeyFlag(col.Flag) && is.Table.PKIsHandle) || col.ID == model.ExtraHandleID {
+					indexCols = append(indexCols, is.dataSourceSchema.Columns[i])
+					setHandle = true
+					break
+				}
+			}
 		}
-		is.NeedCommonHandle = true
-	}
-
-	// If it's double read case, the first index must return handle. So we should add extra handle column
-	// if there isn't a handle column.
-	if isDoubleRead && !setHandle {
-		if !is.Table.IsCommonHandle {
-			indexCols = append(indexCols, &expression.Column{
-				RetType:  types.NewFieldType(mysql.TypeLonglong),
-				ID:       model.ExtraHandleID,
-				UniqueID: is.ctx.GetSessionVars().AllocPlanColumnID(),
-			})
+		// If it's double read case, the first index must return handle. So we should add extra handle column
+		// if there isn't a handle column.
+		if isDoubleRead && !setHandle {
+			if !is.Table.IsCommonHandle {
+				indexCols = append(indexCols, &expression.Column{
+					RetType:  types.NewFieldType(mysql.TypeLonglong),
+					ID:       model.ExtraHandleID,
+					UniqueID: is.ctx.GetSessionVars().AllocPlanColumnID(),
+				})
+			}
+		}
+	} else {
+		if len(is.IdxCols) < len(is.Columns) {
+			for i := len(is.IdxCols); i < len(idxExprCols); i++ {
+				indexCols = append(indexCols, idxExprCols[i])
```

Comment:
Need to check if `idxExprCols[i] != nil`?

Reply:
`idxExprCols` here for common handle never be `nil`.